[ES|QL] CSV Reporting in Discover

Dosant opened the pull request:
## Summary

close https://github.com/elastic/kibana/issues/173390
This PR enables CSV report generation with ES|QL in Discover. 

Pre this PR, there are two report types for generating CSV reports with Discover: 
- https://github.com/elastic/kibana/blob/main/packages/kbn-reporting/export_types/csv/csv_searchsource.ts - old deprecated report type that relies on `SerializedSearchSourceFields`. This type is still currently used by Discover UI and we plan to migrate away from it https://github.com/elastic/kibana/issues/151190
- https://github.com/elastic/kibana/blob/main/packages/kbn-reporting/export_types/csv/csv_v2.ts - new report type that relies on Discover locator. Currently, it can only generate reports from searches backed by saved search and this report type was implemented as a public-facing API for simple report generation outside Kibana. 

Since we plan to migrate to the v2 report type and search source is not needed for es|ql, this PR implements es|ql csv reporting based on the v2 report type.

Initially, I wanted to create new new report type similar to v2 just for es|ql, but it turned out to be a lot more boilerplate code without a significant benefit. you can see it in the PR https://github.com/elastic/kibana/pull/174448. So I changed my mind and the current PR adds es|ql capabilities inside the existing csv_v2 report. This is convenient as the input is the same (Discover locator), the output is the same csv file and meta information, and telemetry is also the same. 

As of this PR, the ES|QL report is capable of: 
- Using es|ql query from the locator 
- Adding time range filter if present in the locator. time field is picked from the data view (it is available in the locator, but otherwise is not used for es|ql). Other filters are also passed if available in the locator 
- Keeps into account "columns" from the locator.
- Similar to current non-es|ql reports from discover UI, it doesn't use saved searches but only relies on the state from the URL. This probably will be improved in https://github.com/elastic/kibana/issues/151190 to support both. 
- Uses different CSV settings for functionality like checking formulas, escapes, bom, max size, etc... 
- Keeps regular CSV features like cancelation and giving event loop a break (even though those are not needed for now for es|ql since the number of results is limited) 


Some notable differences compared to regular discover search / csv reports: 
- A lot simpler, as it doesn't use search source and field formats 
- No pagination, less CPU heavy as esql responses are limited to 10000 results and a single request 


## Questions 

### esql timeouts?

Esql search strategy has custom default timeout logic. But csv reporting has a config for request timeout overrides. Do we need to use the reporting overrides or should we keep default esql timeout logic?

Dosant committed:
esql csv 2

Dosant commented:
/ci

Dosant commented on packages/kbn-es-types/src/search.ts:
```diff
@@ -665,3 +665,13 @@ export interface ESQLSearchReponse {
   values: ESQLRow[];
 }
+
+export interface ESQLSearchParams {
```
this was just moved from common/search/expressions/esql

Dosant commented on packages/kbn-generate-csv/src/generate_csv_esql.ts:
```diff
@@ -0,0 +1,221 @@
+  uiSettings: IUiSettingsClient;
+}
+
+export class CsvESQLGenerator {
```
This is a copy of `CsvGenerator` with changes for es|ql generation. I kept most of the CSV building-specific things like maxSizeBytes, bom, escapeFormulaValues, cancelation, and CPU breaks, and reused the same CSV settings. Inputs/Outputs/errors/warnings are mostly the same

Dosant commented on packages/kbn-generate-csv/src/generate_csv_esql.ts:
```diff
@@ -0,0 +1,221 @@
+        // https://github.com/elastic/elasticsearch/pull/102767
+        // timezone
+        // TODO: pass locale from browser?
+        // locale
```
This wasn't available in the locator, I wasn't eager to add it as wasn't sure if this was significant. Happy to add if needed.

tsullivan replied:
What would `locale` be used for? Is it different than timezone?

Dosant replied:
https://github.com/elastic/kibana/pull/174511#issuecomment-1887037057

tsullivan replied:
Thanks! I wonder if we should not use field formatters for the non-ES|QL CSV export as well. Do you know if any plans are changing with field formatters?

Dosant replied:
> Do you know if any plans are changing with field formatters?
don't know :(

> I wonder if we should not use field formatters for the non-ES|QL CSV export as well.
I thought that the idea was to try to match the Discover table as closely as possible. I think it makes sense to keep the formatters for non-ESQL until they are there in the Discover table.

Dosant commented on src/plugins/discover/public/application/main/components/top_nav/get_top_nav_links.tsx:
```diff
@@ -183,7 +183,9 @@ export const getTopNavLinks = ({
         sharingData: {
-          ...sharingData,
+          isTextBased,
+          locatorParams: [{ id: locator.id, params }],
```
locatorParams will be used for es|ql report (csv_v2), searchSourceSharingData is still used for other reports. 
Hopefully we will only need locatorParams when fully migrated to csv_v2 https://github.com/elastic/kibana/issues/151190

Dosant commented on x-pack/plugins/reporting/public/share_context_menu/reporting_panel_content/reporting_panel_content.tsx:
```diff
@@ -251,7 +251,8 @@ class ReportingPanelContentUi extends Component<Props, State> {
       case CSV_REPORT_TYPE:
-        return 'csv';
+      case CSV_REPORT_TYPE_V2:
+        return 'CSV';
```
fixing https://github.com/elastic/kibana/issues/173837

Dosant commented on packages/kbn-reporting/export_types/csv/csv_v2.ts:
```diff
@@ -87,14 +87,10 @@ export class CsvV2ExportType extends ExportType<
       throw Boom.badRequest('Invalid Job params: must contain a single Discover App locator');
     }
 
-    if (!params || !params.savedSearchId || typeof params.savedSearchId !== 'string') {
```
For the non es|ql reports csv_v2 still needs the savedSearchId. this is asserted deeper inside the code branch. 
 
es|ql reports are working without savedSearchId

full support and alignment for saved and non saved searches should be implemented in https://github.com/elastic/kibana/issues/151190

tsullivan replied:
++ on the idea to implement ES|QL CSV export within the csv_v2 export type.

tsullivan commented on packages/kbn-generate-csv/src/generate_csv_esql.ts:
```diff
@@ -0,0 +1,221 @@
+      const escape = this.escapeValues(settings);
+
+      for (const column of columns) {
+        rowDefinition.push(escape(`${dataTableRow[column]}`));
```
Maybe as you had it was intentional, but this looks the same as:
```suggestion
        rowDefinition.push(escape(dataTableRow[column]));
```

Dosant replied:
safe typecasting from unknown to string + if there is anything something then string, would be safely converted to string

tsullivan commented on x-pack/plugins/reporting/public/share_context_menu/register_csv_reporting.tsx:
```diff
@@ -8,7 +8,7 @@
 import React from 'react';
 
-import { CSV_JOB_TYPE } from '@kbn/reporting-export-types-csv-common';
+import { CSV_JOB_TYPE, CSV_JOB_TYPE_V2 } from '@kbn/reporting-export-types-csv-common';
```
I would just like to give a head's up that this file might be changing quite a bit in https://github.com/elastic/kibana/pull/171253/files

Dosant replied:
😢

Dosant committed:
add test

elasticmachine commented:
Pinging @elastic/appex-sharedux (Team:SharedUX)

stratoula commented:
/ci

stratoula commented:
From my tests so far it works great! About your questions:

- Locale is used as of now by some date functions eg. date_format() to localize the output. I am passing the i18n locale so we could do the same here.

> Esql search strategy has custom default timeout logic. But csv reporting has a config for request timeout overrides. Do we need to use the reporting overrides or should we keep default esql timeout logic?

This is a very good question 🤔 I vote for keeping the default esql timeout logic. @lukasolson wdyt?

A question from my side. We are going to use async query in Discover https://github.com/elastic/kibana/issues/172131. Does it affect the reporting service for ES|QL? And if yes, how?

Dosant committed:
add a test

jloleysens reviewed (approved):
Core changes LGTM!

Dosant commented:
@stratoula, 
> A question from my side. We are going to use async query in Discover https://github.com/elastic/kibana/issues/172131. Does it affect the reporting service for ES|QL? And if yes, how?

Currently, for regular searches, we use async search in Discover, but sync search in reporting. I think this is mostly for historical reasons and it wasn't considered if it is worth switching to async search there, but I am not sure.
In theory, we can switch to async search in reporting and then use reporting-specific timeouts or task cancelation to cancel the query. Still, I am not sure what the benefits of those are compared to using sync search + requestTimeout. 

When migrating to async search for esql in Discover,  we should also discuss if it is worth doing the same for reporting,

stratoula commented:
@Dosant got it. @ppisljar is currently working on it. I hope that we are going to merge it in 8.13. Maybe we need to sync online before?

Dosant committed:
fixes

Dosant committed:
Merge branch 'main' of github.com:elastic/kibana into d/2024-01-09-esql-csv-2

tsullivan commented:
@Dosant @stratoula 

> In theory, we can switch to async search in reporting and then use reporting-specific timeouts or task manager's task cancelation to cancel the query. 

For non-ES|QL CSV export, I investigated using async search (as an alternative to paging) and found it not to be suitable for export use cases, due to there being a size limit of results within Elasticsearch. The size limit would be a hurdle for Reporting since it is common to export multi-gigabytes of data. I'm not sure this would affect ES|QL searches since the size of exported data will be similar to what the user has already seen loaded into their browser.

> This is a very good question 🤔 I vote for keeping the default esql timeout logic. @lukasolson wdyt?

We should assume that some users will use the Reporting APIs directly and craft report job parameters with the number of aggregated terms or the time range filter without testing the query in Kibana first. Therefore, there could be higher pressure on the timeouts for reports vs for queries in the browser. I think we do want those timeout overrides to be available for any kind of report.

tsullivan commented on packages/kbn-reporting/export_types/csv/csv_v2.ts:
```diff
@@ -121,6 +117,36 @@ export class CsvV2ExportType extends ExportType<
+    const query = await locatorClient.queryFromLocator(params);
+
+    if (query && 'esql' in query) {
+      // TODO: use columnsFromLocator
+      // We don't implement it now because it might be a breaking change for csv_v2 api that for non esql only works with saved search
+      // https://github.com/elastic/kibana/issues/151190
+      // const columns = await locatorClient.columnsFromLocator(params);
```
I think I don't follow this. Is it saying there would need to be changes made to the `columnsFromLocator` method in order to have it support esql, and those changes could be breaking for non-esql?

Dosant replied:
> Is it saying there would need to be changes made to the columnsFromLocator method in order to have it support esql, and those changes could be breaking for non-esql?

Yes. I'll try to make the comment clear.


currently locatorClient.columnsFromLocator can only extract columns from the saved search, but for the es|ql I simply wanted to get currently visible columns from params.columns. 
I didn't want to add this change inside locatorClient.columnsFromLocator, as it would change the behaviour of csv_v2 for non-ES|QL export

tsullivan commented on src/plugins/discover/server/locator/query_from_locator.ts:
```diff
@@ -0,0 +1,31 @@
+  ): Promise<Query | AggregateQuery | undefined> => {
+    return params.query;
+
+    // TODO: support query from saved search
```
What would need to happen for this TODO to complete?

Dosant replied:
if would need to take params.savedObjectId, fetch the saved object and take the query from there. 
I didn't do it because we need to decide which query takes precedence (I'd expect the params.query) and this as logic would work differently compared to csv_searchsource I didn't want Discover UI to have different behavior for non-ESQL and ESQL export. I think best would be to clean it up here https://github.com/elastic/kibana/issues/151190

tsullivan commented on x-pack/plugins/reporting/public/share_context_menu/register_csv_reporting.tsx:
```diff
@@ -41,12 +45,21 @@ export const reportingCsvShareProvider = ({
+        };
+      }
+
+      // csv v2 uses search source and columns
```
```suggestion
      // csv v1 uses search source and columns
```

tsullivan commented on x-pack/test/reporting_api_integration/reporting_and_security/__snapshots__/csv_v2_esql.snap:
```diff
@@ -0,0 +1,35 @@
+
+exports[`Reporting APIs CSV Generation from ES|QL export from non-timebased data view csv from es|ql csv file matches 1`] = `
+"eon,epoch,era,period
+Phanerozoic,\\" Pliocene\\",Cenozoic,Neogene
+Phanerozoic,\\" Holocene\\",Cenozoic,Quaternary
```
The spaces before these epoch names are intentional, right?

tsullivan reviewed (commented):
Thanks for getting this going! I have left a few comments and questions for my understanding.

It looks close to approval. I would just like to see where that discussion about timeouts goes before I can approve it.

stratoula commented:
@tsullivan about the timeouts I don't have a strong preference. Whatever makes sense to you. 

About the async query I do understand what you say. There is something we need to take under consideration though. The sync api has a default limit of 500. If I am not mistaken (I will ask the es team to be sure) the async api will raise this constraint and the default will be 10000. So there will be a difference in the behaviors if we use the sync api for reporting and the async api for discover search. I think we could by pass it if we add | limit 10000 in the queries if no limit pipe is detected

stratoula commented:
Update: I discussed it with ES and they told me that the default limit if it changes in the future it will change in both apis so we are good to use the sync api in reporting and async api to fetch the results in Discover 👍

Dosant commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into d/2024-01-09-esql-csv-2

Dosant committed:
fix comments

Dosant committed:
Merge branch 'main' of github.com:elastic/kibana into d/2024-01-09-esql-csv-2

Dosant committed:
Merge branch 'd/2024-01-09-esql-csv-2' of github.com:Dosant/kibana into d/2024-01-09-esql-csv-2

Dosant committed:
use reporting’s timeout

Dosant committed:
fix tests

Dosant commented:
> Thanks for getting this going! I have left a few comments and questions for my understanding.
> 
> It looks close to approval. I would just like to see where that discussion about timeouts goes before I can approve it.

> about the timeouts I don't have a strong preference. Whatever makes sense to you.


@tsullivan, I am now passing the timeout from reporting (it is `scroll.duration` btw which could be a bit confusing, as we don't have a dedicated timeout setting in csv, but this make it aligned with the the non-es|ql search). I am also passing the abort signal so that the search is aborted in case the task manager's task in canceled, I think we should do the same for non-esql. 


@stratoula, I have doubts about this custom timeout. 
```
// Also abort after two mins
    setTimeout(() => abortController.abort(), ES_TIMEOUT_IN_MS);
```

I have updated it, so if a custom requestTimeout is passed, the custom timeout will be used here
But I wonder if this custom timeout is still needed since Lukas fixed bsearch abort here https://github.com/elastic/kibana/pull/169041

stratoula commented:
I think we can remove it now, yes. There was a bug in bsearch and was not aborting correctly but with Lukas PR the bug is fixed. Also we are moving the async api which will make the timeout redundant. @lukasolson can you confirm?

Dosant committed:
remove custom cancelation timoout

it shouldn’t be needed since https://github.com/elastic/kibana/pull/169041 is fixed

tsullivan commented:
@stratoula sounds like today, the sync API and async API have the same document limit, which is great!

@Dosant let's not worry too much about the confusing name of the `scroll.duration` setting being used. There will be a revisit of timeout settings here: https://github.com/elastic/kibana/issues/131852

tsullivan commented on packages/kbn-generate-csv/src/generate_csv_esql.ts:
```diff
@@ -0,0 +1,228 @@
+          IKibanaSearchResponse<ESQLSearchReponse>
+        >(searchParams, {
+          strategy: ESQL_SEARCH_STRATEGY,
+          abortSignal: abortController.signal,
```
Nice!

tsullivan reviewed (approved):
LGTM

stratoula committed:
Merge branch 'main' into d/2024-01-09-esql-csv-2

stratoula commented:
/ci

stratoula reviewed (approved):
LGTM, I tested it locally and works as expected! Thanx @Dosant ❤️

Dosant committed:
add test

Dosant committed:
Merge branch 'main' of github.com:elastic/kibana into d/2024-01-09-esql-csv-2

Dosant committed:
Merge branch 'd/2024-01-09-esql-csv-2' of github.com:Dosant/kibana into d/2024-01-09-esql-csv-2

kibanamachine committed:
Merge branch 'main' into d/2024-01-09-esql-csv-2

jughosta reviewed (commented):
Great work, @Dosant thanks! Adding some question below.

1. I wonder whether it's a temporary solution that we are not using formatters for the ES|QL CVS report? The response from ES request has types for columns which we could use.

2. Noticed that for non-timebased index patterns, the CSV report still highlights the first column. Is it expected? 3. Also, is it expected to have "null" for missing values?

<img width="500" alt="Screenshot 2024-01-16 at 11 55 47" src="https://github.com/elastic/kibana/assets/1415710/2cd2b2a5-f60a-4ce3-9548-71a9ae3064c0">

4. Custom sorting on ES|QL columns in Discover seems to be ignored in the generated CSV reports. Not sure if it's a blocker for this PR or it can be addressed later. cc @stratoula 

Regarding the changes to search strategy, it would be great to get @lukasolson review too.

stratoula commented:
@jughosta 

> I wonder whether it's a temporary solution that we are not using formatters for the ES|QL CVS report? The response from ES request has types for columns which we could use.

We don't use them in Discover table either so I guess this is ok.

> Noticed that for non-timebased index patterns, the CSV report still highlights the first column. Is it expected? 3. Also, is it expected to have "null" for missing values?

Nice catch, it would be great if we could address it

> Custom sorting on ES|QL columns in Discover seems to be ignored in the generated CSV reports. Not sure if it's a blocker for this PR or it can be addressed later. 

We don't save it on the saved search either so I guess we could address it later. I leave this to Anton to decide.

Dosant commented:
@jughosta, thank you!

> I wonder whether it's a temporary solution that we are not using formatters for the ES|QL CVS report? The response from ES request has types for columns which we could use.

Same what @stratoula said 

> Noticed that for non-timebased index patterns, the CSV report still highlights the first column. Is it expected? 

I think the highlighting you mean is just how by default when you opened the CSV in Numbers apps it applied the template where the first column is highlighted. Is this what you mean? The csv itself doesn't have any formatting. 

> Also, is it expected to have "null" for missing values?

Agree, I'll try to improve it now 

> Custom sorting on ES|QL columns in Discover seems to be ignored in the generated CSV reports. Not sure if it's a blocker for this PR or it can be addressed later. 
> We don't save it on the saved search either so I guess we could address it later. I leave this to Anton to decide.

Right, the thing is that it isn't part of the state or locator or a saved search, I didn't want to change it for this pr. But I don't think this is that important in an CSV as it can be done with one click in Excel/Sheets

stratoula commented:
> I think the highlighting you mean is just how by default when you opened the CSV in Numbers apps it applied the template where the first column is highlighted. Is this what you mean? The csv itself doesn't have any formatting.

Aaa yes Anton! I am also always confused with this but this is a numbers "feature" :D

Dosant committed:
improve empty value handling

Dosant committed:
Merge branch 'd/2024-01-09-esql-csv-2' of github.com:Dosant/kibana into d/2024-01-09-esql-csv-2

Dosant commented:
/ci

jughosta reviewed (approved):
Thanks! LGTM 👍

lukasolson reviewed (approved):
Search changes LGTM

Dosant commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into d/2024-01-09-esql-csv-2
